### PR TITLE
Add warnings about products availability

### PR DESCRIPTION
**Pull request by jeffersontuc**
#### What problem is this solving?

`shipping-calculator` needs to know the amount of unavailable items as well as the total amount of items, since we want to show some proper warnings about the availability of the products to a given address. This PR does it.

#### How should this be manually tested?

- Proceed to [Workspace](https://deliveryfeedback--checkoutio.myvtex.com/cart/add?sku=308&sku=288)
- Verify that a warning (image 1) it's being shown.
- Remove the available item.
- Verify that another type of warning (image 2) is being shown now.

#### Screenshots or example usage

**Image 1**

image

**Image 2**

image

#### Checklist/Reminders

- [ ] Updated `README.md`.
- [x] Updated `CHANGELOG.md`.
- [x] [Linked this PR to a Clubhouse story (if applicable).](https://app.clubhouse.io/vtex/story/27145/implementar-tratamento-de-erros-para-delivery-options)
- [ ] Updated/created tests (important for bug fixes).
- [ ] Deleted the workspace after merging this PR (if applicable).

#### Type of changes

✔️ | Type of Change
---|---
_ | Bug fix 
✔️| New feature 
_ | Breaking change 
_ | Technical improvements

### CODE REVIEW

File: messages/en.json [hunk 1]
```diff
@@ -4,6 +4,9 @@
   "store/shipping-calculator.edit": "edit",
   "store/shipping-calculator.estimate": "estimate",
   "store/shipping-calculator.viewDeliveryOptions": "View delivery options",
+  "store/shipping-calculator.allItemsUnavailable": "All products can't be delivered to this address.",
+  "store/shipping-calculator.unavailableItems": "{numberOfUnavailableItems, plural, one {# product} other {# products}} that can't be delivered to this address are highlighted.",
```
Comment (marcoskwkm) [inline messages/en.json]: You need to pluralize `is/are` too.

File: messages/en.json [hunk 2]
```diff
@@ -4,6 +4,9 @@
...
+  "store/shipping-calculator.optionsForRemainingProducts": "{numberOfOptions, plural, one {Option} other {Options}} for remaining {numberOfItems, plural, one {product} other {products}}",
```
Comment (marcoskwkm) [inline messages/en.json:9]: You only treated plural for this message in `en`. Is this intentional?

File: messages/pt.json [hunk 3]
```diff
@@ -4,6 +4,9 @@
   "store/shipping-calculator.edit": "alterar",
   "store/shipping-calculator.estimate": "calcular",
   "store/shipping-calculator.viewDeliveryOptions": "Ver opções de entrega",
+  "store/shipping-calculator.allItemsUnavailable": "Nenhum dos produtos pode ser entregue nesse endereço.",
+  "store/shipping-calculator.unavailableItems": "Atenção a {numberOfUnavailableItems, plural, one {# produto que não pode ser entregue} other {# produtos que não podem ser entregues}} neste endereço.",
+  "store/shipping-calculator.optionsForRemainingProducts": "Opções para os produtos restantes",
```
Comment (marcoskwkm) [inline messages/pt.json]: "Atenção a um produto que não pode ser entregue neste endereço" is a bit different from the `en` message. Could you confirm with @davicosta99 what would be the message in Portuguese?

File: messages/pt.json [hunk 4]
```diff
@@ -4,6 +4,9 @@
   "store/shipping-calculator.edit": "alterar",
   "store/shipping-calculator.estimate": "calcular",
   "store/shipping-calculator.viewDeliveryOptions": "Ver opções de entrega",
+  "store/shipping-calculator.allItemsUnavailable": "Nenhum dos produtos pode ser entregue nesse endereço.",
+  "store/shipping-calculator.unavailableItems": "Atenção a {numberOfUnavailableItems, plural, one {# produto que não pode ser entregue} other {# produtos que não podem ser entregues}} neste endereço.",
```
Comment (darlenedms) [inline messages/pt.json]:
```suggestion
  "store/shipping-calculator.unavailableItems": "Atenção, há {numberOfUnavailableItems, plural, one {# produto que não pode ser entregue} other {# produtos que não podem ser entregues}} neste endereço.",
```

File: messages/es.json [hunk 5]
```diff
@@ -4,6 +4,9 @@
   "store/shipping-calculator.edit": "cambiar",
   "store/shipping-calculator.estimate": "calcular",
   "store/shipping-calculator.viewDeliveryOptions": "Ver opciones de entrega",
+  "store/shipping-calculator.allItemsUnavailable": "Ningún de los productos se puede entregar a esta dirección.",
+  "store/shipping-calculator.unavailableItems": "Atención a {numberOfUnavailableItems, plural, one {# producto que no se puede} other {# productos que no se pueden}} entregar a esta dirección.",
```
Comment (darlenedms) [inline messages/es.json]:
```suggestion
  "store/shipping-calculator.unavailableItems": "Atención, hay {numberOfUnavailableItems, plural, one {# producto que no se puede} other {# productos que no se pueden}} entregar a esta dirección.",
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+
+- Warning to when there are unavailable products.
+- Warning to when all the products are unavailable.
```
Comment (marcoskwkm) [inline CHANGELOG.md]: Remove `to`.

**Review by marcoskwkm — CHANGES_REQUESTED** [review]

File: messages/pt.json (hunk 3 above)
Comment (jeffersontuc) [reply messages/pt.json]: It was @augustocb who told me the `pt-br` translation, actually.

File: messages/pt.json (hunk 4 above)
Comment (jeffersontuc) [reply messages/pt.json]: I think that what @augustocb wanted to say is "to pay attention TO the products", but he can say better.

File: react/components/ShippingOptions.tsx [hunk 7]
```diff
@@ -18,6 +24,22 @@ const ShippingOptions: FunctionComponent<CustomProps> = ({
 
   return (
     <Fragment>
+      <DeliveryOptionsAvailability
+        numberOfItems={numberOfItems}
+        numberOfUnavailableItems={numberOfUnavailableItems}
+      />
+      {numberOfUnavailableItems ? (
```
Comment (darlenedms) [inline react/components/ShippingOptions.tsx]: In general we use the `&&` without returning `null` in this scenarios.

File: react/components/DeliveryOptionsAvailability.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+const DeliveryOptionsAvailability: FunctionComponent<Props> = ({
+  numberOfItems,
+  numberOfUnavailableItems,
+}) => {
+  const allItemsUnavailable = numberOfUnavailableItems === numberOfItems
+  return (
+    <Fragment>
+      {numberOfUnavailableItems ? (
```
Comment (darlenedms) [inline react/components/DeliveryOptionsAvailability.tsx]: In general we use the `&&` without returning `null` in this scenarios.

**Review by darlenedms — CHANGES_REQUESTED** [review]
Messages and a JSX pattern we use.

File: messages/en.json (hunk 2 above)
Comment (jeffersontuc) [reply messages/en.json:9]: It's not. thanks!

**Comment by darlenedms** [comment]
I know it's not directly related to this PR but I would like @davicosta99 and @augustocb attention on "remove" button hover state in product list's warning message. Is that correct?
Screen Shot 2019-12-16 at 4 23 22 PM

File: react/components/ShippingOptions.tsx (hunk 7 above)
Comment (jeffersontuc) [reply react/components/ShippingOptions.tsx]: I was doing that before, but when `numberOfUnavailableItems` was zero, it was showing the zero and we do not want it.

File: react/components/DeliveryOptionsAvailability.tsx (hunk 8 above)
Comment (jeffersontuc) [reply react/components/DeliveryOptionsAvailability.tsx]: ditto

File: react/components/ShippingOptions.tsx (hunk 7 above)
Comment (marcoskwkm) [reply react/components/ShippingOptions.tsx]: What about `numberOfUnavailableItems > 0 &&`?

File: messages/pt.json (hunk 3 above)
Comment (jeffersontuc) [reply messages/pt.json]: We just discussed this but we couldn't came to a conclusion. We are waiting for @davicosta99 to say what he thinks.

File: react/components/ShippingOptions.tsx (hunk 7 above)
Comment (jeffersontuc) [reply react/components/ShippingOptions.tsx]: that works. nice!

Comment (jeffersontuc) [reply react/components/ShippingOptions.tsx]: that works. thanks @darlenedms @marcoskwkm !

**Comment by davicosta99** [comment]
> I know it's not directly related to this PR but I would like @davicosta99 and @augustocb attention on "remove" button hover state in product list's warning message. Is that correct?
> Screen Shot 2019-12-16 at 4 23 22 PM

teste

**Comment by davicostalf** [comment]
> I know it's not directly related to this PR but I would like @davicosta99 and @augustocb attention on "remove" button hover state in product list's warning message. Is that correct?
> Screen Shot 2019-12-16 at 4 23 22 PM

@augustocb just pointed out that the "remove" label has an ambiguous message that may be interpreted   as hiding the message. Since there's already the trash can, better to remove the button for now as we explore other options.

Regarding the hover, it's a issue for all styleguide buttons in colored backgrounds. We have to design a solution for it in the design language level.

File: messages/en.json (hunk 2 above)
Comment (davicostalf) [reply messages/en.json:9]: First message should be: "All the products can't be delivered to this address."
Sorry for the change 😉

**Review by marcoskwkm — DISMISSED** [review]

File: messages/pt.json (hunk 4 above)
Comment (davicostalf) [reply messages/pt.json]: "Nenhum dos produtos pode**m** ser entregue nesse endereço".

"Podem" instead of "pode".

Comment (jeffersontuc) [reply messages/pt.json]: In this case, I think it's also "entregues" instead of "entregue", right? @davicosta99

File: messages/es.json (hunk 5 above)
Comment (jeffersontuc) [reply messages/es.json]: @augustocb @davicosta99 we need to decide this also.

File: messages/pt.json (hunk 4 above)
Comment (davicostalf) [reply messages/pt.json]: All the messages:

**EN**
"This product can't be delivered to this address"
"All the products can't be delivered to this address."

"A product that can't be delivered to this address was highlighted."
"# products that can't be delivered to this address were highlighted."

**PT-BR**
"Esse produto não pode ser entregue nesse endereço."
"Nenhum dos produtos podem ser entregues nesse endereço."

"Um produto que não pode ser entregue nesse endereço foi destacado."
"# produtos que não podem ser entregues nesse endereço foram destacados."

**ES**
"Este producto no se puede entregar en esta dirección."
"Ningún de los productos se puede entregar en esta dirección."

"Se encuentra resaltado un producto que no se puede entregar en esta dirección."
"Se encuentran marcados # productos que no se pueden entregar en esta dirección."

-----
Sorry for the confusion, I'll make sure to include these translations along with the UI in the future.

**Review by marcoskwkm — DISMISSED** [review]

**Review by davicostalf — COMMENTED** [review]
The label "Opção para o produto restante" is being shown with no options below it. Is this related to our test environment?

File: messages/pt.json (hunk 4 above)
Comment (davicostalf) [reply messages/pt.json]: Final fix (I hope):

"The product can't be delivered to this address"
"O produto não pode ser entregue nesse endereço."
"El producto no se puede entregar en esta dirección."

**Comment by davicosta99** [comment]
Great. Approved by me.

**Comment by jeffersontuc** [comment]
> The label "Opção para o produto restante" is being shown with no options below it. Is this related to our test environment?

No. This is related to our graphql, and it was already implemented, by the way.

**Review by darlenedms — APPROVED** [review]

**Review by marcoskwkm — APPROVED** [review]

**Pull request merged**